File: bindings/swift/Sources/SQLiteVec/SQLiteVec.swift
import Foundation
import SQLite
import CSQLiteVec
import SQLite3

/// Swift bindings for the sqlite-vec SQLite extension
public class SQLiteVec {
    
    /// Load sqlite-vec into a SQLite.swift database connection
    /// - Parameter database: The SQLite.swift database connection
    /// - Throws: Database.Error if loading fails
    public static func load(database: Connection) throws {
        // Get the raw sqlite3 handle from SQLite.swift
        let handle = database.handle
        
        // Initialize sqlite-vec directly using the C function
        var errorMessage: UnsafeMutablePointer<CChar>? = nil
        let result = sqlite3_vec_init(handle, &errorMessage, nil)
        
        if result != SQLITE_OK {
            let error = errorMessage != nil ? String(cString: errorMessage!) : "Unknown error loading sqlite-vec"
            if let errorMessage = errorMessage {
                sqlite3_free(errorMessage)
            }
            throw NSError(domain: "SQLiteVec", code: Int(result), userInfo: [NSLocalizedDescriptionKey: error])
        }
    }
    
    /// Initialize sqlite-vec for auto-loading in all future database connections
    /// Note: Auto-extension is deprecated on Apple platforms. Consider using manual loading instead.
    public static func autoLoad() {
        #if !os(iOS) && !os(macOS) && !os(watchOS) && !os(tvOS)
        let autoExtensionFunction: @convention(c) (OpaquePointer?, UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?, UnsafePointer<sqlite3_api_routines>?) -> Int32 = { db, pzErrMsg, pApi in
            return sqlite3_vec_init(db, pzErrMsg, pApi)
        }
        sqlite3_auto_extension(unsafeBitCast(autoExtensionFunction, to: (@convention(c) () -> Void).self))
        #else
        print("Warning: Auto-extension is deprecated on Apple platforms. Use manual loading with SQLiteVec.load(database:) instead.")
        #endif
    }
    
    /// Cancel auto-loading of sqlite-vec for future database connections
    /// Note: Auto-extension is deprecated on Apple platforms.
    public static func cancelAutoLoad() {
        #if !os(iOS) && !os(macOS) && !os(watchOS) && !os(tvOS)
        let autoExtensionFunction: @convention(c) (OpaquePointer?, UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?, UnsafePointer<sqlite3_api_routines>?) -> Int32 = { db, pzErrMsg, pApi in
            return sqlite3_vec_init(db, pzErrMsg, pApi)
        }
        sqlite3_cancel_auto_extension(unsafeBitCast(autoExtensionFunction, to: (@convention(c) () -> Void).self))
        #else
        print("Warning: Auto-extension cancellation is not supported on Apple platforms.")
        #endif
    }
    
    /// Get the version of sqlite-vec
    /// - Parameter database: The SQLite.swift database connection
    /// - Returns: The version string
    /// - Throws: Database.Error if the query fails
    public static func version(database: Connection) throws -> String {
        let result = try database.scalar("SELECT vec_version()") as! String
        return result
    }
}

// MARK: - Vector Serialization

extension SQLiteVec {
    
    /// Serialize a Float32 array into raw bytes format that sqlite-vec expects
    /// - Parameter vector: Array of Float32 values
    /// - Returns: Data containing the serialized vector
    public static func serialize(_ vector: [Float32]) -> Data {
        return Data(bytes: vector, count: vector.count * MemoryLayout<Float32>.size)
    }
    
    /// Serialize a Float64 array into raw bytes format that sqlite-vec expects
    /// - Parameter vector: Array of Float64 values  
    /// - Returns: Data containing the serialized vector
    public static func serialize(_ vector: [Float64]) -> Data {
        return Data(bytes: vector, count: vector.count * MemoryLayout<Float64>.size)
    }
    
    /// Serialize an Int8 array into raw bytes format that sqlite-vec expects
    /// - Parameter vector: Array of Int8 values
    /// - Returns: Data containing the serialized vector
    public static func serialize(_ vector: [Int8]) -> Data {
        return Data(bytes: vector, count: vector.count * MemoryLayout<Int8>.size)
    }
    
    /// Serialize an array of floats into raw bytes format
    /// - Parameter vector: Array of float values
    /// - Returns: Data containing the serialized vector
    public static func serializeFloat32(_ vector: [Float]) -> Data {
        let float32Vector = vector.map { Float32($0) }
        return serialize(float32Vector)
    }
    
    /// Deserialize raw bytes into a Float32 array
    /// - Parameter data: Raw bytes from sqlite-vec
    /// - Returns: Array of Float32 values
    public static func deserializeFloat32(_ data: Data) -> [Float32] {
        return data.withUnsafeBytes { bytes in
            let buffer = bytes.bindMemory(to: Float32.self)
            return Array(buffer)
        }
    }
    
    /// Deserialize raw bytes into an Int8 array
    /// - Parameter data: Raw bytes from sqlite-vec
    /// - Returns: Array of Int8 values
    public static func deserializeInt8(_ data: Data) -> [Int8] {
        return data.withUnsafeBytes { bytes in
            let buffer = bytes.bindMemory(to: Int8.self)
            return Array(buffer)
        }
    }
}

// MARK: - Convenience Extensions

extension Connection {
    
    /// Load the sqlite-vec extension into this database connection
    /// - Throws: Database.Error if loading fails
    public func loadSQLiteVec() throws {
        try SQLiteVec.load(database: self)
    }
}

// MARK: - Vector Operations

extension SQLiteVec {
    
    /// Calculate the distance between two vectors using sqlite-vec
    /// - Parameters:
    ///   - database: The SQLite.swift database connection
    ///   - vector1: First vector as Float32 array
    ///   - vector2: Second vector as Float32 array
    ///   - metric: Distance metric (default: "cosine")
    /// - Returns: The calculated distance
    /// - Throws: Database.Error if the query fails
    public static func distance(
        database: Connection,
        vector1: [Float32],
        vector2: [Float32],
        metric: String = "cosine"
    ) throws -> Double {
        let data1 = serialize(vector1)
        let data2 = serialize(vector2)
        
        let query: String
        switch metric.lowercased() {
        case "cosine":
            query = "SELECT vec_distance_cosine(?, ?)"
        case "l2":
            query = "SELECT vec_distance_l2(?, ?)"
        default:
            query = "SELECT vec_distance_cosine(?, ?)"
        }
        
        let result = try database.scalar(query, [Blob(bytes: [UInt8](data1)), Blob(bytes: [UInt8](data2))]) as! Double
        return result
    }
    
    /// Get the length (number of dimensions) of a vector
    /// - Parameters:
    ///   - database: The SQLite.swift database connection
    ///   - vector: Vector as Float32 array
    /// - Returns: The number of dimensions
    /// - Throws: Database.Error if the query fails
    public static func length(database: Connection, vector: [Float32]) throws -> Int {
        let data = serialize(vector)
        let result = try database.scalar("SELECT vec_length(?)", [Blob(bytes: [UInt8](data))]) as! Int64
        return Int(result)
    }
    
    /// Normalize a vector using sqlite-vec
    /// - Parameters:
    ///   - database: The SQLite.swift database connection
    ///   - vector: Vector as Float32 array to normalize
    /// - Returns: Normalized vector as Float32 array
    /// - Throws: Database.Error if the query fails
    public static func normalize(database: Connection, vector: [Float32]) throws -> [Float32] {
        let data = serialize(vector)
        let resultData = try database.scalar("SELECT vec_normalize(?)", [Blob(bytes: [UInt8](data))]) as! Blob
        return deserializeFloat32(Data(resultData.bytes))
    }
}

// MARK: - Vector Table Operations

extension SQLiteVec {
    
    /// Create a vec0 virtual table
    /// - Parameters:
    ///   - database: The SQLite.swift database connection
    ///   - tableName: Name of the table to create
    ///   - vectorColumn: Name and definition of the vector column (e.g., "embedding float[384]")
    ///   - additionalColumns: Additional column definitions
    /// - Throws: Database.Error if the creation fails
    public static func createVectorTable(
        database: Connection,
        tableName: String,
        vectorColumn: String,
        additionalColumns: [String] = []
    ) throws {
        var columns = [vectorColumn]
        columns.append(contentsOf: additionalColumns)
        let columnDefinitions = columns.joined(separator: ", ")
        
        let sql = "CREATE VIRTUAL TABLE \(tableName) USING vec0(\(columnDefinitions))"
        try database.execute(sql)
    }
    
    /// Search for similar vectors in a vec0 table
    /// - Parameters:
    ///   - database: The SQLite.swift database connection
    ///   - tableName: Name of the vec0 table
    ///   - vectorColumn: Name of the vector column
    ///   - queryVector: Vector to search for similarities
    ///   - limit: Maximum number of results (default: 10)
    ///   - additionalColumns: Additional columns to select
    /// - Returns: Array of search results
    /// - Throws: Database.Error if the search fails
    public static func searchSimilar(
        database: Connection,
        tableName: String,
        vectorColumn: String,
        queryVector: [Float32],
        limit: Int = 10,
        additionalColumns: [String] = []
    ) throws -> [[Any?]] {
        let queryData = serialize(queryVector)
        
        var selectColumns = ["rowid", "distance"]
        selectColumns.append(contentsOf: additionalColumns)
        let selectClause = selectColumns.joined(separator: ", ")
        
        let sql = """
            SELECT \(selectClause)
            FROM \(tableName)
            WHERE \(vectorColumn) MATCH ?
            ORDER BY distance
            LIMIT ?
        """
        
        var results: [[Any?]] = []
        for row in try database.prepare(sql, [Blob(bytes: [UInt8](queryData)), limit]) {
            var rowData: [Any?] = []
            for i in 0..<selectColumns.count {
                rowData.append(row[i])
            }
            results.append(rowData)
        }
        
        return results
    }
}

// MARK: - Additional Utility Functions (Go bindings compatibility)

extension SQLiteVec {
    
    /// Serialize a vector array for Int8 vectors similar to Go bindings
    /// - Parameter vector: Array of Int8 values  
    /// - Returns: Data containing the serialized vector
    public static func serializeInt8(_ vector: [Int8]) -> Data {
        return serialize(vector)
    }
    
    /// Check if sqlite-vec is available in the database
    /// - Parameter database: The SQLite.swift database connection
    /// - Returns: True if sqlite-vec is loaded and available
    public static func isAvailable(database: Connection) -> Bool {
        do {
            _ = try version(database: database)
            return true
        } catch {
            return false
        }
    }
    
    /// Load sqlite-vec if not already loaded
    /// - Parameter database: The SQLite.swift database connection
    /// - Throws: Database.Error if loading fails
    public static func ensureLoaded(database: Connection) throws {
        if !isAvailable(database: database) {
            try load(database: database)
        }
    }
}

// MARK: - Vector Math Utilities

extension SQLiteVec {
    
    /// Calculate dot product between two vectors using sqlite-vec
    /// - Parameters:
    ///   - database: The SQLite.swift database connection
    ///   - vector1: First vector as Float32 array
    ///   - vector2: Second vector as Float32 array
    /// - Returns: The dot product result
    /// - Throws: Database.Error if the query fails
    public static func dotProduct(
        database: Connection,
        vector1: [Float32],
        vector2: [Float32]
    ) throws -> Double {
        let data1 = serialize(vector1)
        let data2 = serialize(vector2)
        
        let result = try database.scalar("SELECT vec_distance_cosine(?, ?)", [Blob(bytes: [UInt8](data1)), Blob(bytes: [UInt8](data2))]) as! Double
        return result
    }
    
    /// Add two vectors using sqlite-vec
    /// - Parameters:
    ///   - database: The SQLite.swift database connection
    ///   - vector1: First vector as Float32 array
    ///   - vector2: Second vector as Float32 array
    /// - Returns: The resulting vector as Float32 array
    /// - Throws: Database.Error if the query fails
    public static func add(
        database: Connection,
        vector1: [Float32],
        vector2: [Float32]
    ) throws -> [Float32] {
        let data1 = serialize(vector1)
        let data2 = serialize(vector2)
        
        let resultData = try database.scalar("SELECT vec_add(?, ?)", [Blob(bytes: [UInt8](data1)), Blob(bytes: [UInt8](data2))]) as! Blob
        return deserializeFloat32(Data(resultData.bytes))
    }
    
    /// Subtract two vectors using sqlite-vec
    /// - Parameters:
    ///   - database: The SQLite.swift database connection
    ///   - vector1: First vector as Float32 array
    ///   - vector2: Second vector as Float32 array
    /// - Returns: The resulting vector as Float32 array
    /// - Throws: Database.Error if the query fails
    public static func subtract(
        database: Connection,
        vector1: [Float32],
        vector2: [Float32]
    ) throws -> [Float32] {
        let data1 = serialize(vector1)
        let data2 = serialize(vector2)
        
        let resultData = try database.scalar("SELECT vec_sub(?, ?)", [Blob(bytes: [UInt8](data1)), Blob(bytes: [UInt8](data2))]) as! Blob
        return deserializeFloat32(Data(resultData.bytes))
    }
    
    /// Convert a vector to JSON representation using sqlite-vec
    /// - Parameters:
    ///   - database: The SQLite.swift database connection
    ///   - vector: Vector as Float32 array
    /// - Returns: JSON string representation of the vector
    /// - Throws: Database.Error if the query fails
    public static func toJSON(database: Connection, vector: [Float32]) throws -> String {
        let data = serialize(vector)
        let result = try database.scalar("SELECT vec_to_json(?)", [Blob(bytes: [UInt8](data))]) as! String
        return result
    }
}

// MARK: - Batch Operations (Go-style)

extension SQLiteVec {
    
    /// Insert multiple vectors in a batch operation for better performance
    /// Similar to Go bindings pattern for bulk inserts
    /// - Parameters:
    ///   - database: The SQLite.swift database connection
    ///   - tableName: Name of the vec0 table
    ///   - vectorColumn: Name of the vector column
    ///   - vectors: Array of tuples containing (id, vector) pairs
    /// - Throws: Database.Error if the operation fails
    public static func insertVectorsBatch(
        database: Connection,
        tableName: String,
        vectorColumn: String,
        vectors: [(Int64, [Float32])]
    ) throws {
        try database.transaction {
            let sql = "INSERT INTO \(tableName)(rowid, \(vectorColumn)) VALUES (?, ?)"
            for (id, vector) in vectors {
                let vectorData = serialize(vector)
                try database.run(sql, id, Blob(bytes: [UInt8](vectorData)))
            }
        }
    }
    
    /// Search similar vectors with additional filter conditions
    /// More flexible version similar to Go bindings approach
    /// - Parameters:
    ///   - database: The SQLite.swift database connection
    ///   - tableName: Name of the vec0 table
    ///   - vectorColumn: Name of the vector column
    ///   - queryVector: Vector to search for similarities
    ///   - limit: Maximum number of results
    ///   - whereClause: Additional WHERE conditions (optional)
    ///   - parameters: Parameters for the WHERE clause
    /// - Returns: Array of search results
    /// - Throws: Database.Error if the search fails
    public static func searchSimilarWithFilter(
        database: Connection,
        tableName: String,
        vectorColumn: String,
        queryVector: [Float32],
        limit: Int = 10,
        whereClause: String? = nil,
        parameters: [Binding?] = []
    ) throws -> [[Any?]] {
        let queryData = serialize(queryVector)
        
        var sql = """
            SELECT rowid, distance
            FROM \(tableName)
            WHERE \(vectorColumn) MATCH ?
        """
        
        var allParameters: [Binding?] = [Blob(bytes: [UInt8](queryData))]
        
        if let whereClause = whereClause {
            sql += " AND (\(whereClause))"
            allParameters.append(contentsOf: parameters)
        }
        
        sql += " ORDER BY distance LIMIT ?"
        allParameters.append(limit)
        
        var results: [[Any?]] = []
        for row in try database.prepare(sql, allParameters) {
            results.append([row[0], row[1]])
        }
        
        return results
    }
}

File: bindings/android/src/main/kotlin/com/sqlite/vec/SQLiteVec.kt
package com.sqlite.vec

import android.content.Context
import android.database.Cursor
import android.database.sqlite.SQLiteDatabase
import android.database.sqlite.SQLiteException
import android.database.sqlite.SQLiteOpenHelper
import android.util.Log
import java.nio.ByteBuffer
import java.nio.ByteOrder

/**
 * Android bindings for sqlite-vec - Vector search for SQLite
 * 
 * This class provides Kotlin/Java bindings for the sqlite-vec SQLite extension,
 * enabling vector search capabilities in Android SQLite databases.
 * 
 * Features:
 * - Fast vector search powered by sqlite-vec
 * - Support for Float32, Float64, and Int8 vectors
 * - Cross-platform compatibility across all Android architectures
 * - Easy integration with existing Android SQLite code
 * - Comprehensive error handling and logging
 * 
 * Example usage:
 * ```kotlin
 * val db = SQLiteDatabase.openDatabase(":memory:", null, SQLiteDatabase.OPEN_READWRITE)
 * SQLiteVec.load(db)
 * 
 * val version = SQLiteVec.version(db)
 * println("sqlite-vec version: $version")
 * 
 * // Create vector table
 * SQLiteVec.createVectorTable(db, "documents", "embedding float[384]")
 * 
 * // Insert vectors
 * val embedding = floatArrayOf(0.1f, 0.2f, 0.3f, 0.4f)
 * val vectorData = SQLiteVec.serialize(embedding)
 * db.execSQL("INSERT INTO documents(rowid, embedding) VALUES (?, ?)", arrayOf(1, vectorData))
 * 
 * // Search similar vectors
 * val queryVector = floatArrayOf(0.2f, 0.2f, 0.3f, 0.3f)
 * val results = SQLiteVec.searchSimilar(db, "documents", "embedding", queryVector, 5)
 * ```
 */
class SQLiteVec {
    
    companion object {
        private const val TAG = "SQLiteVec"
        private var isLoaded = false
        private var autoLoadEnabled = false
        
        init {
            try {
                System.loadLibrary("sqlite-vec")
                isLoaded = true
                Log.i(TAG, "sqlite-vec native library loaded successfully")
            } catch (e: UnsatisfiedLinkError) {
                Log.e(TAG, "Failed to load sqlite-vec native library", e)
                isLoaded = false
            }
        }
        
        // Native method declarations
        @JvmStatic
        private external fun nativeGetVersion(): String
        
        @JvmStatic
        private external fun nativeInit(dbHandle: Long): Int
        
        @JvmStatic
        private external fun nativeSerializeFloat32(vector: FloatArray): ByteArray?
        
        @JvmStatic
        private external fun nativeDeserializeFloat32(data: ByteArray): FloatArray?
        
        @JvmStatic
        private external fun nativeSerializeInt8(vector: ByteArray): ByteArray?
        
        @JvmStatic
        private external fun nativeDistance(dbHandle: Long, vector1: ByteArray, vector2: ByteArray, metric: String): Double
        
        @JvmStatic
        private external fun nativeVectorLength(dbHandle: Long, vector: ByteArray): Int
        
        @JvmStatic
        private external fun nativeNormalize(dbHandle: Long, vector: ByteArray): ByteArray?
        
        /**
         * Load sqlite-vec extension into a SQLite database
         * 
         * @param database The SQLite database to load the extension into
         * @throws SQLiteException if loading fails
         */
        @JvmStatic
        fun load(database: SQLiteDatabase) {
            if (!isLoaded) {
                throw SQLiteException("sqlite-vec native library not loaded")
            }
            
            try {
                val handle = getDatabaseHandle(database)
                val result = nativeInit(handle)
                if (result != 0) { // SQLITE_OK = 0
                    throw SQLiteException("Failed to initialize sqlite-vec (error code: $result)")
                }
                Log.i(TAG, "sqlite-vec extension loaded successfully")
            } catch (e: Exception) {
                Log.e(TAG, "Failed to load sqlite-vec extension", e)
                throw SQLiteException("Failed to load sqlite-vec extension: ${e.message}", e)
            }
        }
        
        /**
         * Get the version of the sqlite-vec extension
         * 
         * @return Version string (e.g., "v0.1.7-alpha.2")
         */
        @JvmStatic
        fun getVersion(): String {
            if (!isLoaded) {
                throw SQLiteException("sqlite-vec native library not loaded")
            }
            return nativeGetVersion()
        }
        
        /**
         * Get the version of sqlite-vec from a loaded database
         * 
         * @param database The SQLite database with sqlite-vec loaded
         * @return Version string from the database
         * @throws SQLiteException if query fails
         */
        @JvmStatic
        fun version(database: SQLiteDatabase): String {
            return database.rawQuery("SELECT vec_version()", null).use { cursor ->
                if (cursor.moveToFirst()) {
                    cursor.getString(0)
                } else {
                    throw SQLiteException("Failed to get sqlite-vec version")
                }
            }
        }
        
        /**
         * Check if sqlite-vec is available in the database
         * 
         * @param database The SQLite database to check
         * @return true if sqlite-vec is loaded and available
         */
        @JvmStatic
        fun isAvailable(database: SQLiteDatabase): Boolean {
            return try {
                version(database)
                true
            } catch (e: Exception) {
                false
            }
        }
        
        /**
         * Load sqlite-vec if not already loaded
         * 
         * @param database The SQLite database
         * @throws SQLiteException if loading fails
         */
        @JvmStatic
        fun ensureLoaded(database: SQLiteDatabase) {
            if (!isAvailable(database)) {
                load(database)
            }
        }
        
        /**
         * Enable auto-loading of sqlite-vec for future database connections
         * Note: This is a placeholder - actual implementation would require
         * hooking into SQLiteOpenHelper or database factory
         */
        @JvmStatic
        fun autoLoad() {
            autoLoadEnabled = true
            Log.i(TAG, "sqlite-vec auto-loading enabled")
        }
        
        /**
         * Disable auto-loading of sqlite-vec
         */
        @JvmStatic
        fun cancelAutoLoad() {
            autoLoadEnabled = false
            Log.i(TAG, "sqlite-vec auto-loading disabled")
        }
        
        /**
         * Serialize a Float32 array into raw bytes format that sqlite-vec expects
         * 
         * @param vector Array of Float values
         * @return ByteArray containing the serialized vector
         */
        @JvmStatic
        fun serialize(vector: FloatArray): ByteArray {
            if (!isLoaded) {
                throw SQLiteException("sqlite-vec native library not loaded")
            }
            return nativeSerializeFloat32(vector) 
                ?: throw SQLiteException("Failed to serialize Float32 vector")
        }
        
        /**
         * Serialize a Double array into raw bytes format
         * 
         * @param vector Array of Double values
         * @return ByteArray containing the serialized vector
         */
        @JvmStatic
        fun serialize(vector: DoubleArray): ByteArray {
            val buffer = ByteBuffer.allocate(vector.size * 8)
            buffer.order(ByteOrder.nativeOrder())
            for (value in vector) {
                buffer.putDouble(value)
            }
            return buffer.array()
        }
        
        /**
         * Serialize an Int8 array (ByteArray) into raw bytes format
         * 
         * @param vector Array of Byte values
         * @return ByteArray containing the serialized vector
         */
        @JvmStatic
        fun serialize(vector: ByteArray): ByteArray {
            if (!isLoaded) {
                throw SQLiteException("sqlite-vec native library not loaded")
            }
            return nativeSerializeInt8(vector) 
                ?: throw SQLiteException("Failed to serialize Int8 vector")
        }
        
        /**
         * Serialize a Float32 array (Go bindings compatibility)
         * 
         * @param vector Array of Float values
         * @return ByteArray containing the serialized vector
         */
        @JvmStatic
        fun serializeFloat32(vector: FloatArray): ByteArray = serialize(vector)
        
        /**
         * Serialize an Int8 array (Go bindings compatibility)
         * 
         * @param vector Array of Byte values
         * @return ByteArray containing the serialized vector
         */
        @JvmStatic
        fun serializeInt8(vector: ByteArray): ByteArray = serialize(vector)
        
        /**
         * Deserialize raw bytes into a Float32 array
         * 
         * @param data Raw bytes from sqlite-vec
         * @return FloatArray of deserialized values
         */
        @JvmStatic
        fun deserializeFloat32(data: ByteArray): FloatArray {
            if (!isLoaded) {
                throw SQLiteException("sqlite-vec native library not loaded")
            }
            return nativeDeserializeFloat32(data) 
                ?: throw SQLiteException("Failed to deserialize Float32 vector")
        }
        
        /**
         * Deserialize raw bytes into an Int8 array
         * 
         * @param data Raw bytes from sqlite-vec
         * @return ByteArray of deserialized values
         */
        @JvmStatic
        fun deserializeInt8(data: ByteArray): ByteArray {
            // For Int8, no conversion needed
            return data.clone()
        }
        
        /**
         * Calculate the distance between two vectors using sqlite-vec
         * 
         * @param database The SQLite database with sqlite-vec loaded
         * @param vector1 First vector as FloatArray
         * @param vector2 Second vector as FloatArray
         * @param metric Distance metric ("cosine" or "l2", default: "cosine")
         * @return The calculated distance
         * @throws SQLiteException if calculation fails
         */
        @JvmStatic
        fun distance(
            database: SQLiteDatabase,
            vector1: FloatArray,
            vector2: FloatArray,
            metric: String = "cosine"
        ): Double {
            if (!isLoaded) {
                throw SQLiteException("sqlite-vec native library not loaded")
            }
            
            val data1 = serialize(vector1)
            val data2 = serialize(vector2)
            val handle = getDatabaseHandle(database)
            
            val result = nativeDistance(handle, data1, data2, metric)
            if (result < 0) {
                throw SQLiteException("Failed to calculate vector distance")
            }
            return result
        }
        
        /**
         * Get the length (number of dimensions) of a vector
         * 
         * @param database The SQLite database with sqlite-vec loaded
         * @param vector Vector as FloatArray
         * @return The number of dimensions
         * @throws SQLiteException if query fails
         */
        @JvmStatic
        fun length(database: SQLiteDatabase, vector: FloatArray): Int {
            if (!isLoaded) {
                throw SQLiteException("sqlite-vec native library not loaded")
            }
            
            val data = serialize(vector)
            val handle = getDatabaseHandle(database)
            
            val result = nativeVectorLength(handle, data)
            if (result < 0) {
                throw SQLiteException("Failed to get vector length")
            }
            return result
        }
        
        /**
         * Normalize a vector using sqlite-vec
         * 
         * @param database The SQLite database with sqlite-vec loaded
         * @param vector Vector as FloatArray to normalize
         * @return Normalized vector as FloatArray
         * @throws SQLiteException if normalization fails
         */
        @JvmStatic
        fun normalize(database: SQLiteDatabase, vector: FloatArray): FloatArray {
            if (!isLoaded) {
                throw SQLiteException("sqlite-vec native library not loaded")
            }
            
            val data = serialize(vector)
            val handle = getDatabaseHandle(database)
            
            val resultData = nativeNormalize(handle, data)
                ?: throw SQLiteException("Failed to normalize vector")
                
            return deserializeFloat32(resultData)
        }
        
        /**
         * Create a vec0 virtual table
         * 
         * @param database The SQLite database with sqlite-vec loaded
         * @param tableName Name of the table to create
         * @param vectorColumn Name and definition of the vector column (e.g., "embedding float[384]")
         * @param additionalColumns Additional column definitions
         * @throws SQLiteException if table creation fails
         */
        @JvmStatic
        fun createVectorTable(
            database: SQLiteDatabase,
            tableName: String,
            vectorColumn: String,
            vararg additionalColumns: String
        ) {
            val columns = mutableListOf(vectorColumn)
            columns.addAll(additionalColumns)
            val columnDefinitions = columns.joinToString(", ")
            
            val sql = "CREATE VIRTUAL TABLE $tableName USING vec0($columnDefinitions)"
            try {
                database.execSQL(sql)
                Log.d(TAG, "Created vector table: $tableName")
            } catch (e: Exception) {
                Log.e(TAG, "Failed to create vector table: $tableName", e)
                throw SQLiteException("Failed to create vector table: $tableName", e)
            }
        }
        
        /**
         * Search for similar vectors in a vec0 table
         * 
         * @param database The SQLite database with sqlite-vec loaded
         * @param tableName Name of the vec0 table
         * @param vectorColumn Name of the vector column
         * @param queryVector Vector to search for similarities
         * @param limit Maximum number of results (default: 10)
         * @param additionalColumns Additional columns to select
         * @return List of search results as Maps
         * @throws SQLiteException if search fails
         */
        @JvmStatic
        fun searchSimilar(
            database: SQLiteDatabase,
            tableName: String,
            vectorColumn: String,
            queryVector: FloatArray,
            limit: Int = 10,
            vararg additionalColumns: String
        ): List<Map<String, Any?>> {
            val queryData = serialize(queryVector)
            
            val selectColumns = mutableListOf("rowid", "distance")
            selectColumns.addAll(additionalColumns)
            val selectClause = selectColumns.joinToString(", ")
            
            val sql = """
                SELECT $selectClause
                FROM $tableName
                WHERE $vectorColumn MATCH ?
                ORDER BY distance
                LIMIT ?
            """.trimIndent()
            
            val results = mutableListOf<Map<String, Any?>>()
            
            try {
                database.rawQuery(sql, arrayOf(queryData, limit.toString())).use { cursor ->
                    while (cursor.moveToNext()) {
                        val row = mutableMapOf<String, Any?>()
                        for (i in 0 until cursor.columnCount) {
                            val columnName = cursor.getColumnName(i)
                            val value = when (cursor.getType(i)) {
                                Cursor.FIELD_TYPE_INTEGER -> cursor.getLong(i)
                                Cursor.FIELD_TYPE_FLOAT -> cursor.getDouble(i)
                                Cursor.FIELD_TYPE_STRING -> cursor.getString(i)
                                Cursor.FIELD_TYPE_BLOB -> cursor.getBlob(i)
                                else -> null
                            }
                            row[columnName] = value
                        }
                        results.add(row)
                    }
                }
            } catch (e: Exception) {
                Log.e(TAG, "Failed to search similar vectors", e)
                throw SQLiteException("Failed to search similar vectors", e)
            }
            
            return results
        }
        
        /**
         * Convert a vector to JSON representation using sqlite-vec
         * 
         * @param database The SQLite database with sqlite-vec loaded
         * @param vector Vector as FloatArray
         * @return JSON string representation of the vector
         * @throws SQLiteException if query fails
         */
        @JvmStatic
        fun toJSON(database: SQLiteDatabase, vector: FloatArray): String {
            val data = serialize(vector)
            
            return database.rawQuery("SELECT vec_to_json(?)", arrayOf(data)).use { cursor ->
                if (cursor.moveToFirst()) {
                    cursor.getString(0)
                } else {
                    throw SQLiteException("Failed to convert vector to JSON")
                }
            }
        }
        
        /**
         * Insert multiple vectors in a batch operation for better performance
         * Similar to Swift bindings pattern for bulk inserts
         * 
         * @param database The SQLite database with sqlite-vec loaded
         * @param tableName Name of the vec0 table
         * @param vectorColumn Name of the vector column
         * @param vectors List of pairs containing (id, vector)
         * @throws SQLiteException if the operation fails
         */
        @JvmStatic
        fun insertVectorsBatch(
            database: SQLiteDatabase,
            tableName: String,
            vectorColumn: String,
            vectors: List<Pair<Long, FloatArray>>
        ) {
            if (vectors.isEmpty()) return
            
            val sql = "INSERT INTO $tableName(rowid, $vectorColumn) VALUES (?, ?)"
            
            try {
                database.beginTransaction()
                
                for ((id, vector) in vectors) {
                    val vectorData = serialize(vector)
                    database.execSQL(sql, arrayOf(id, vectorData))
                }
                
                database.setTransactionSuccessful()
                Log.d(TAG, "Successfully inserted ${vectors.size} vectors")
                
            } catch (e: Exception) {
                Log.e(TAG, "Failed to insert vectors batch", e)
                throw SQLiteException("Failed to insert vectors batch", e)
            } finally {
                database.endTransaction()
            }
        }
        
        /**
         * Helper function to get database handle from SQLiteDatabase
         * This uses reflection as Android doesn't expose the native handle directly
         */
        private fun getDatabaseHandle(database: SQLiteDatabase): Long {
            try {
                // Try different field names based on Android version
                val handleFields = listOf("mNativeHandle", "mConnectionPtr", "mConnection")
                
                for (fieldName in handleFields) {
                    try {
                        val field = database.javaClass.getDeclaredField(fieldName)
                        field.isAccessible = true
                        val handle = field.getLong(database)
                        if (handle != 0L) {
                            return handle
                        }
                    } catch (e: NoSuchFieldException) {
                        continue
                    }
                }
                
                throw SQLiteException("Could not access database handle")
            } catch (e: Exception) {
                Log.e(TAG, "Failed to get database handle", e)
                throw SQLiteException("Failed to get database handle", e)
            }
        }
    }
}
